apps/evans/components/core/EmbeddedAudio.tsx

```diff
@@ -1,60 +1,176 @@
-import React from 'react';
+import React, { useCallback, useEffect, useRef, useState } from 'react';
 import cx from 'classnames';
-import { LANG } from '@/lib/env';
+import { BackwardIcon, ForwardIcon } from '@heroicons/react/24/solid';
+import styles from '../../styles/AudioPlayer.module.css';
+import { formatDuration } from '@/lib/dates';
 
 interface Props {
-  className?: string;
-  title: string;
-  trackId: number;
-  playlistId?: number | null;
-  showArtwork?: boolean;
-  height?: number;
+  tracks: Array<{
+    title: string;
+    url: string;
+  }>;
 }
 
-const EmbeddedAudio: React.FC<Props> = ({
-  className = ``,
-  trackId,
-  playlistId,
-  showArtwork = true,
-  height = 166,
-  title,
-}) => (
-  <iframe
-    className={cx(`bg-flgray-100`, className)}
-    title={title}
-    width="100%"
-    height={String(height)}
-    scrolling="no"
-    frameBorder="no"
-    src={src(trackId, playlistId, showArtwork)}
-  />
-);
+const EmbeddedAudio: React.FC<Props> = ({ tracks }) => {
+  const [isPlaying, setIsPlaying] = useState(false);
+  const [currentTime, setCurrentTime] = useState(0);
 
-export default EmbeddedAudio;
+  const currentTrack = useRef(0);
+  const audioRef = useRef<HTMLAudioElement>(null);
+  const animationRef = useRef<number>();
+  const [audio, setAudio] = useState<HTMLAudioElement | null>(null);
 
-function src(
-  trackId: number,
-  playlistId: number | undefined | null,
-  showArtwork: boolean,
-): string {
-  const resource = playlistId ? `playlists` : `tracks`;
-  const id = playlistId || trackId;
-  const params: Record<string, string> = {
-    url: `https://api.soundcloud.com/${resource}/${id}`,
-    color: LANG === `es` ? `C18C59` : `6C3142`,
-    auto_play: `false`,
-    hide_related: `true`,
-    show_comments: `false`,
-    show_user: `false`,
-    show_reposts: `false`,
-    show_artwork: showArtwork ? `true` : `false`,
-    show_teaser: `false`,
-    locale: LANG,
-  };
-
-  const query = Object.keys(params)
-    .map((key) => `${encodeURIComponent(key)}=${encodeURIComponent(params[key] ?? ``)}`)
-    .join(`&`);
-
-  return `https://w.soundcloud.com/player/?${query}`;
-}
+  const currentTitle = tracks[currentTrack.current]?.title ?? ``;
+
+  const whilePlaying = useCallback(() => {
+    if (!audio) return;
+    if (audio.currentTime === audio.duration) {
```

Comment:
i might be misunderstanding this, but it seems like we're only running this function to progress to the next track, or stop if we're at the end of the last track.  is that right? but we're running the function on every animation frame? i'm not super experienced with requestanimationframe, so maybe this is normal practice, but it seems sort of wasteful to query 60 times per seconds if we're at the end...  is there another way to do this? or maybe it doens't matter. just thinking aloud

Reply:
Yeah you're totally right, just removed the callback on every frame and it totally work great. Thanks!